src/lib/x509/certstor_system_windows/certstor_windows.cpp

```diff
@@ -185,6 +188,52 @@ std::vector<X509_Certificate> find_cert_by_dn_and_key_id(const Botan::X509_DN& s
 
    return search_cert_stores(blob, find_type, filter, return_on_first_found);
    }
+
+/** Handle certificates that do not adhere to RFC 3280 using a subject key identifier
+ *  that is not equal to the SHA-1 of the public key (w/o algorithm identifier)
+ *
+ *  This method lazily builds a cache of certificates found in previous queries as well
+ *  as negative results for @p key_hash queries that didn't find a certificate.
+ *
+ *  See here for further details: https://github.com/randombit/botan/issues/2779
+ */
+std::optional<X509_Certificate> find_cert_by_pubkey_sha1_via_exhaustive_search(const std::vector<uint8_t> &key_hash)
+   {
+   static std::map<std::vector<uint8_t>, std::optional<X509_Certificate>> non_rfc3289_certs;
```

Comment:
It seems like there could be concurrency issues here, if one created multiple certificate stores and performed queries in different threads. Should this instead be a member variable?

Reply:
True. My intention was to share the cache with multiple instances of the certificate store to accomodate people that re-instanciate it for every lookup. Nevertheless, the potential race is definitely a fair point.

I changed it to a member variable, with one caveat: All methods of this class are `const`-qualified: I.e. I was forced to qualify the member variable as `mutable`. This might cause usage-dependent threading issues as well, as a user might feel safe to use a `const`-only class without locking.

Would you go for a mutex in that situation, to play it safe?